gguf-py/gguf/constants.py

```diff
@@ -936,6 +935,366 @@ def get_type(val: Any) -> GGUFValueType:
+BPE_PRE_TOKENIZER_DEFAULT = "'s|'t|'re|'ve|'m|'ll|'d| ?\\p{L}+| ?\\p{N}+| ?[^\\s\\p{L}\\p{N}]+|\\s+(?!\\S)"
+GPT_PRE_TOKENIZER_DEFAULT = f"{BPE_PRE_TOKENIZER_DEFAULT}|\\s+"
+
+# Default pre-processing regex for BPE tokenization
+BPE_PRE_PROCESSOR_DEFAULT = (
+    "[\\p{P}\\$\\+<=>\\^~\\|]+", BPE_PRE_TOKENIZER_DEFAULT, "\\p{N}+", "[0-9][0-9][0-9]",
+)
+
+# NOTE: GPT-2 is the standard default pre-tokenizer for all models
+# NOTE: BERT models inherit from the Byte Level Pre-tokenizer.
+# https://github.com/huggingface/tokenizers/blob/main/tokenizers/src/pre_tokenizers/byte_level.rs#L117
+# https://github.com/huggingface/tokenizers/blob/main/tokenizers/src/pre_tokenizers/bert.rs#L13
+BPE_PRE_TOKENIZERS = {
+    # gpt2, olmo, phi (1, 1_5, 2, 3, ...)
+    "gpt2": (GPT_PRE_TOKENIZER_DEFAULT,),
+    # dbrx
+    # NOTE: PR#6920: https://github.com/ggerganov/llama.cpp/pull/6920#issuecomment-2080233989
+    "llama3": (
+        "(?:'[sS]|'[tT]|'[rR][eE]|'[vV][eE]|'[mM]|'[lL][lL]|'[dD])|[^\\r\\n\\p{L}\\p{N}]?\\p{L}+|\\p{N}{1,3}| ?[^\\s\\p{L}\\p{N}]+[\\r\\n]*|\\s*[\\r\\n]+|\\s+(?!\\S)|\\s+",
+    ),
+    "falcon": (
+        "[\\p{P}\\$\\+<=>\\^~\\|]+", BPE_PRE_TOKENIZER_DEFAULT, "[0-9][0-9][0-9]",
+    ),
+    "mpt": (
+        "\\s?\\p{L}+", "\\s?\\p{P}+", BPE_PRE_TOKENIZER_DEFAULT,
+    ),
+    # starcoder, refact, command-r
+    "starcoder": (
+        "\\p{N}", BPE_PRE_TOKENIZER_DEFAULT,
+    ),
+    # qwen, qwen2, stablelm
+    "qwen": (
+        "(?:'[sS]|'[tT]|'[rR][eE]|'[vV][eE]|'[mM]|'[lL][lL]|'[dD])|[^\\r\\n\\p{L}\\p{N}]?\\p{L}+|\\p{N}| ?[^\\s\\p{L}\\p{N}]+[\\r\\n]*|\\s*[\\r\\n]+|\\s+(?!\\S)|\\s+",
+    ),
+    # NOTE: deepseek uses the 'llama' arch, but diverges with the pre-tok.
+    "deepseek": (
+        "[\r\n]",
+        "\\s?[A-Za-zµÀ-ÖØ-öø-ƺƼ-ƿǄ-ʓʕ-ʯͰ-ͳͶͷͻ-ͽͿΆΈ-ΊΌΎ-ΡΣ-ϵϷ-ҁҊ-ԯԱ-ՖႠ-ჅᎠ-Ᏽᏸ-ᏽᲐ-ᲺᲽ-Ჿᴀ-ᴫᵫ-ᵷᵹ-ᶚḀ-ἕἘ-Ἕἠ-ὅὈ-Ὅὐ-ὗὙὛὝὟ-ώᾀ-ᾴᾶ-ᾼιῂ-ῄῆ-ῌῐ-ΐῖ-Ίῠ-Ῥῲ-ῴῶ-ῼℂℇℊ-ℓℕℙ-ℝℤΩℨK-ℭℯ-ℴℹℼ-ℿⅅ-ⅉⅎↃↄⰀ-ⱻⱾ-ⳤⳫ-ⳮⳲⳳꙀ-ꙭꚀ-ꚛꜢ-ꝯꝱ-ꞇꞋ-ꞎꭰ-ꮿﬀ-ﬆﬓ-ﬗＡ-Ｚａ-ｚ𐐀-𐑏𐒰-𐓓𐓘-𐓻𐲀-𐲲𐳀-𐳲𑢠-𑣟𞤀-𞥃]+",
+        "\\s?[!-/:-~！-／：-～‘-‟　-。]+",
+        "\\s+$",
+        "[一-龥ࠀ-一가-퟿]+",
+        "\\p{N}+",
+    ),
+    "deepseek-coder": (
+        "[\r\n]",
+        "\\s?\\p{L}+",
+        "\\s?\\p{P}+",
+        "[一-龥ࠀ-一가-퟿]+",
+        "\\p{N}",
+    ),
+    # NOTE: ONLY ADD MODELS ON A AS NEEDED BASIS.
+    # This will get out of control if not properly managed.
+    # This needs a proper solution. The short-term solution is to manually build a map here.
+    # A proper long-term solution would be to build a dynamic registry.
+    # The issue is that this requires a dynamically persistent mapping or a database.
+    # Possible solutions are to use JSON, HDF5, or SQLite.
+    # Some of these mappings could be dynamically generated, but it's sketchy at best.
+    # Model versions should be included along with the model name to mitigate name conflicts.
+    # This entire setup is extremely fragile and will result in breaking changes in the future.
```

Comment:
Can't the `pre_tokenizer` from `tokenizer.json` be used to build the regex of each model instead? There could probably be some handling of all `pre_tokenizer` types (including `Digits`, `Split`, `ByteLevel`, and others) to turn them into a regex or something else that `llama.cpp` could use to achieve the exact same behavior as HuggingFace's `tokenziers` for any `tokenizer.json`?

Is there a specific reason not to go this way and instead manually listing everything?

Reply:
@compilade I tried to do that months ago. It's a no go unfortunately. Some models have the regex in there and others don't.

We need it in advance which is why I'm presuming @ggerganov implemented it the way he did. I'm actively attempting to figure it out, but I think this is more on the model creator than anything.

It would be nice if `tokenizers` baked it in by default, but for some reason these details are often missing. I'm assuming it has something to do with the implementation details on how the hyperparameters are set before training takes place.

Reply:
I'm thinking at some point in the future I can implement a custom BPE tokenizer independently and then integrate it with llama.cpp. Not sure if that's ideal, we'll see.

Reply:
> Some models have the regex in there and others don't.

Apart from GPT-2, which has a "standard" regex hard-coded in HuggingFace's `tokenizers`, I would be very curious of an example of a model which omits necessary information from `tokenizer.json`. I'll try to check each model listed then I'll edit this with a link to the first "broken" `tokenizer.json` I find.

I think if `llama.cpp`'s pretokenizing behavior is made to be exactly the same as `tokenizers`, since `transformers` is usually the suggested way by model creators to run their models, then any fixes by the model creators to `tokenizers.json` would directly apply to `llama.cpp` conversions. Hard-coding what is already in the model files doesn't feel right.

`tokenizers` isn't doing magic, it's reading what it can from `tokenizer.json`. If that file doesn't contain the necessary stuff, then there are built-in defaults which I think should be used by `llama.cpp` too.

If a clean workaround isn't found by then, I'll try to implement this in July. But I have other things I want to finish implementing before then (Jamba, RWKV, etc.).